File: frontend/src/types/rbac.ts
export interface PolicyRule {
  api_groups: string[]
  resources: string[]
  verbs: string[]
  resource_names?: string[]
}

export interface RoleRead {
  name: string
  namespace?: string
  rules: PolicyRule[]
  is_system: boolean
}

export type SubjectKind = 'User' | 'Group' | 'ServiceAccount'

export interface Subject {
  kind: SubjectKind
  name: string
  namespace?: string
}

export interface BindingRead {
  name: string
  namespace?: string
  role_ref: string
  role_kind: string
  subjects: Subject[]
}

export interface UserPermissionSummary {
  username: string
  cluster_bindings: BindingRead[]
  namespace_bindings: BindingRead[]
}

export interface AssignRolePayload {
  role_name: string
  role_kind: 'ClusterRole' | 'Role'
  namespace?: string
}

export interface NamespaceAccessEntry {
  subject: string
  subject_kind: 'User' | 'Group' | 'ServiceAccount'
  subject_namespace?: string
  role: string
  role_kind: 'Role' | 'ClusterRole'
  binding: string
  scope: 'namespace' | 'cluster'
}

export interface CheckAccessRequest {
  user: string
  verb: string
  resource: string
  namespace?: string
  api_group?: string
}

export interface CheckAccessResult {
  allowed: boolean
  denied: boolean
  reason: string
}

File: backend/app/models/rbac.py
from enum import StrEnum
from typing import Generic, TypeVar

from pydantic import BaseModel, Field

T = TypeVar("T")


class PaginatedList(BaseModel, Generic[T]):
    items: list[T]
    total: int
    next_continue: str | None = None


class PolicyRule(BaseModel):
    api_groups: list[str] = Field(default_factory=lambda: [""])
    resources: list[str]
    verbs: list[str]
    resource_names: list[str] | None = None


class RoleCreate(BaseModel):
    name: str
    namespace: str
    rules: list[PolicyRule]


class ClusterRoleCreate(BaseModel):
    name: str
    rules: list[PolicyRule]


class RoleUpdate(BaseModel):
    rules: list[PolicyRule]


class RoleRead(BaseModel):
    name: str
    namespace: str | None = None
    rules: list[PolicyRule]
    is_system: bool = False


class SubjectKind(StrEnum):
    User = "User"
    Group = "Group"
    ServiceAccount = "ServiceAccount"


class Subject(BaseModel):
    kind: SubjectKind
    name: str
    namespace: str | None = None


class BindingCreate(BaseModel):
    name: str
    role_name: str
    role_kind: str = Field(..., pattern=r"^(ClusterRole|Role)$")
    subjects: list[Subject]
    namespace: str | None = None


class BindingRead(BaseModel):
    name: str
    namespace: str | None
    role_ref: str
    role_kind: str
    subjects: list[Subject]


class AssignRoleRequest(BaseModel):
    role_name: str
    role_kind: str = Field(default="ClusterRole", pattern=r"^(ClusterRole|Role)$")
    namespace: str | None = None


class UserPermissionSummary(BaseModel):
    username: str
    cluster_bindings: list[BindingRead]
    namespace_bindings: list[BindingRead]


class NamespaceAccessEntry(BaseModel):
    subject: str
    subject_kind: str
    subject_namespace: str | None = None
    role: str
    role_kind: str
    binding: str
    scope: str  # 'namespace' | 'cluster'


class CheckAccessRequest(BaseModel):
    user: str
    verb: str
    resource: str
    namespace: str | None = None
    api_group: str = ""


class CheckAccessResult(BaseModel):
    allowed: bool
    denied: bool
    reason: str
